core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -1542,4 +1542,50 @@ package object config {
     .bytesConf(ByteUnit.BYTE)
     .createOptional
 
+  private[spark] val GRACEFUL_DECOMMISSION_ENABLE =
+    ConfigBuilder("spark.graceful.decommission.enable")
+      .doc("Whether to enable the node graceful decommissioning handling")
+      .booleanConf
+      .createWithDefault(false)
+
+  private[spark] val GRACEFUL_DECOMMISSION_FETCHFAILED_IGNORE_THRESHOLD =
+    ConfigBuilder("spark.graceful.decommission.fetchfailed.ignore.threshold")
+      .doc("Threshold of number of times fetchfailed ignored due to node" +
+        " decommission.This is configurable as per the need of the user and" +
+        " depending upon type of the cloud. If we keep this a large value and " +
+        " there is continuous decommission of nodes, in those scenarios stage" +
+        " will never abort and keeps on retrying in an unbounded manner.")
+      .intConf
+      .createWithDefault(8)
+
+  private[spark] val GRACEFUL_DECOMMISSION_EXECUTOR_LEASETIME_PCT =
+    ConfigBuilder("spark.graceful.decommission.executor.leasetimePct")
+      .doc("Percentage of time to expiry after which executors are killed " +
+        "(if enabled) on the node. Value ranges between (0-100)")
+      .intConf
+      .checkValue(v => v >= 0 && v < 100, "The percentage should be positive.")
+      .createWithDefault(50) // Pulled out of thin air.
+
+  private[spark] val GRACEFUL_DECOMMISSION_SHUFFLEDATA_LEASETIME_PCT =
+    ConfigBuilder("spark.graceful.decommission.shuffedata.leasetimePct")
+      .doc("Percentage of time to expiry after which shuffle data " +
+        "cleaned up (if enabled) on the node. Value ranges between (0-100)")
+      .intConf
+      .checkValue(v => v >= 0 && v < 100, "The percentage should be positive.")
+      .createWithDefault(90) // Pulled out of thin air.
+
+  private[spark] val GRACEFUL_DECOMMISSION_MIN_TERMINATION_TIME_IN_SEC =
+    ConfigBuilder("spark.graceful.decommission.min.termination.time")
+      .doc("Minimum time to termination below which node decommissioning is performed immediately")
+      .timeConf(TimeUnit.SECONDS)
+      .createWithDefaultString("60s")
+
+  private[spark] val GRACEFUL_DECOMMISSION_NODE_TIMEOUT =
+    ConfigBuilder("spark.graceful.decommission.node.timeout")
```

Comment:
I'm not sure I get the spot loss interaction here, the AM/RM won't know I think and we'd see more of like the K8s one happen.

Reply:
We can get the decommission timeout for hadoop-3.1 and later version of hadoop, so we can use that value to decide when the node is decommissioned.
Whereas for lower version of hadoop(hadoop-2.8) there is no decommissionTimeout  for decommissioning nodes in those scenario we already knew from our experience that in AWS  spotloss nodes will stay for 2 min and GCP preemptible VM will stay for 30 sec after receiving the node decommissioning from hadoop end. 

This config is added here to make decommissioning of nodes to work with multiple version of hadoop, 

 Please find the logic used in YarnAllocator.scala to decide the timeout of the node

```
if (x.getNodeState.toString.equals(NodeState.DECOMMISSIONING.toString)) {
          // In hadoop 2.7 there is no support getDecommissioningTimeout whereas
          // In hadoop 3.1 and later version of hadoop there is support
          // of getDecommissioningTimeout So the method call made using reflection
          // to update the value nodeTerminationTime and for lower version of hadoop2.7
          // use the config spark.graceful.decommission.node.timeout which is specific to cloud
          var nodeTerminationTime = clock.getTimeMillis() + nodeLossInterval * 1000
          try {
              val decommiossioningTimeout = x.getClass.getMethod(
                "getDecommissioningTimeout").invoke(x).asInstanceOf[Integer]
              if (decommiossioningTimeout != null) {
                nodeTerminationTime = clock.getTimeMillis() + decommiossioningTimeout * 1000
              }
          } catch {
            case e: NoSuchMethodException => logDebug(e.toString)
          }
```

Reply:
so what happens with this in hadoop 3.1 and greater? is it ignored if a timeout is specified from yarn?

Reply:
For hadoop3.1 and later version of hadoop there is an interface to get the value of decommissioning timeout method name -  getDecommissioningTimeout. So we have used that here in the code 

```
val decommiossioningTimeout = x.getClass.getMethod(
                "getDecommissioningTimeout").invoke(x).asInstanceOf[Integer]
              if (decommiossioningTimeout != null) {
                nodeTerminationTime = clock.getTimeMillis() + decommiossioningTimeout * 1000
              }
```

Since we are getting the value of decommiossioningTimeout  from the RM in that scenario we will be using that value otherwise we use the value specified in the config.

And also if someone has backported the hadoop3.1 change to lower version of hadoop2.8 etc , For them also they can use decommiossioningTimeout instead of using config GRACEFUL_DECOMMISSION_NODE_TIMEOUT

Reply:
ok so we need to update description to say when it applies

Reply:
Updated the description